internal/mode/static/config_updater.go

```diff
@@ -0,0 +1,105 @@
+package static
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/tools/record"
+
+	nkgAPI "github.com/nginxinc/nginx-kubernetes-gateway/apis/v1alpha1"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/framework/helpers"
+)
+
+// ZapLogLevelSetter defines an interface for setting the logging level of a zap logger.
+type ZapLogLevelSetter interface {
+	SetLevel(string) error
+	Enabled(zapcore.Level) bool
+}
+
+type zapSetterImpl struct {
+	atomicLevel zap.AtomicLevel
+}
+
+func newZapLogLevelSetter(atomicLevel zap.AtomicLevel) zapSetterImpl {
+	return zapSetterImpl{
+		atomicLevel: atomicLevel,
+	}
+}
+
+// SetLevel sets the logging level for the zap logger.
+func (z zapSetterImpl) SetLevel(level string) error {
+	parsedLevel, err := zapcore.ParseLevel(level)
+	if err != nil {
+		fieldErr := field.NotSupported(
+			field.NewPath("logging.level"),
+			level,
+			[]string{
+				string(nkgAPI.ControllerLogLevelInfo),
+				string(nkgAPI.ControllerLogLevelDebug),
+				string(nkgAPI.ControllerLogLevelError),
+			})
+		return fieldErr
+	}
+	z.atomicLevel.SetLevel(parsedLevel)
+
+	return nil
+}
+
+// Enabled returns true if the given level is at or above the current level.
+func (z zapSetterImpl) Enabled(level zapcore.Level) bool {
+	return z.atomicLevel.Enabled(level)
+}
+
+// updateControlPlane updates the control plane configuration with the given user spec.
+// If any fields are not set within the user spec, the default configuration values are used.
+func updateControlPlane(
+	cfg *nkgAPI.NginxGateway,
+	logger logr.Logger,
+	eventRecorder record.EventRecorder,
+	configNSName types.NamespacedName,
+	logLevelSetter ZapLogLevelSetter,
+) error {
+	// build up default configuration
+	controlConfig := nkgAPI.NginxGatewaySpec{
+		Logging: &nkgAPI.Logging{
+			Level: helpers.GetPointer(nkgAPI.ControllerLogLevelInfo),
+		},
+	}
+
+	// by marshaling the user config and then unmarshaling on top of the default config,
+	// we ensure that any unset user values are set with the default values
+	if cfg != nil {
+		cfgBytes, err := json.Marshal(cfg.Spec)
```

Comment:
I'd rather not panic here. If we panic, everything stops. If we log an error, NKG can at least continue doing other things while this issue can be diagnosed.

It's one thing to panic on startup, it's a whole other thing to panic during regular processing.

Reply:
if it is a programmer error, better to panic for consistency, since it is per our guidelines -- https://github.com/nginxinc/nginx-kubernetes-gateway/blob/main/docs/developer/go-style-guide.md#use-panics-for-unrecoverable-errors-or-programming-errors

Reply:
Technically json.Marshal/Unmarshal can fail due to bad field values. Unlikely to happen due to validation beforehand, but not necessarily a programmer error always in this case.